### PR TITLE
chore(renovate): enable PRs for all non-kubernetes dependencies

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

Simplified config to address #4127.

We initially decided to only have security-related updates to dependencies, see https://github.com/crossplane/crossplane/pull/3575#discussion_r1072854569, but that complicates a lot dealing with indirect dependencies (see #4127), I feel like we now have been using renovate enough to switch back to opening PRs for all updates only on master, we can always close the PRs if we don't think we need them, but at least we'll know about the updates and we can always reopen them if needed. We'd still keep the recently introduced rule to disable kubernetes related updates, as those should come in as part of `crossplane-runtime` version bumps first.

We shouldn't fear regressions being introduced, we have a strong suite of unit tests and work is being done to improve also on the E2Es, #4101.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Deployed in my fork, see https://github.com/phisco/crossplane/issues/137 for the PRs that would be created at the moment.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: .github/renovate.json5 [hunk 1]
```diff
@@ -20,18 +21,15 @@
   "labels": ["automated"],
   "regexManagers": [
     {
-// We want a PR to bump Go versions used through env variables in any Github
-// Actions, taking it from the official Github repository.
+      "description": "Bump Go version ued in workflows",
```
Comment (HonkingGoose) [inline .github/renovate.json5:24]:
```suggestion
      "description": "Bump Go version used in workflows",
```

File: .github/renovate.json5 [hunk 2]
```diff
@@ -66,23 +60,26 @@
...
+// be at the beginning, high priority at the end
   "packageRules": [
-// We don't want dependency updates which are not security related for
-// release branches, except for docker images
     {
+      "description": "Ignore non-security related updates to release branches",
       matchBaseBranches: [ "/^release-.*/"],
       enabled: false,
     }, {
+      "description": "Still update docker images on release branches though",
```
Comment (HonkingGoose) [inline .github/renovate.json5:72]:
```suggestion
      "description": "Still update Docker images on release branches though",
```

File: .github/renovate.json5 [hunk 3]
```diff
@@ -66,23 +60,26 @@
...
       "matchDatasources": ["docker"],
       matchBaseBranches: [ "/^release-.*/"],
       enabled: true,
     }, {
+      "description": "Only get docker image updates every 2 weeks to reduce noise",
```
Comment (HonkingGoose) [inline .github/renovate.json5:77]:
```suggestion
      "description": "Only get Docker image updates every 2 weeks to reduce noise",
```

**Review by HonkingGoose — COMMENTED** [review]
I see you're using the `helpers:pinGitHubActionDigests` preset, and use `description` fields, like I suggested. 😄 

There's a typo, and the proper noun `Docker` should be capitalized.

**Comment by negz** [comment]
@phisco If it were possible for Renovate to catch when we needed to update an indirect dependency due to a known vulnerability I would still prefer that. Given that doesn't seem possible, I mostly feel okay with this change.

I'm not really worried about regressions, or increased maintenance work due to extra Renovate PRs to review. My only concern is (or was) that semi-automatically updating all of our dependencies whenever a new version was available could negate some of the benefits of Go's minimal version selection. I re-read https://research.swtch.com/vgo-principles today to refresh my memory on what those benefits are, and to see whether my concern was founded.

I believe the short version is that given a hypothetical diamond dependency:

```mermaid
  graph TD;
      crossplane-->controller-runtime;
      crossplane-->client-go;
      controller-runtime-->yaml;
      client-go-->yaml;
```

Where:

* The latest available `yaml == 1.5`
* `controller-runtime` requires `yaml <= 1.0`
* `client-go` requires `yaml <= 1.1`

Most package managers would pick (and lock) `yaml == 1.5`. Go would instead pick `yaml == 1.1` - i.e. the oldest possible version that satisfies all constraints. It chooses to do this because:

* It enables reproducible builds without using a lock file (which is hard to make work for libraries).
* It lets you avoid dealing with incompatible diamond dependencies until you actually need to. (e.g. To address a vulnerability, or get access to new functionality.)

I don't think semi-automatically updating the minimum versions we depend on hurts reproducible builds in any way. The nature of minimal version selection ensures any particular commit of Crossplane (or something depending on a particular commit of  crossplane-runtime) will be reproducible, dependencies wise.

It does seem like semi-automatically updating the minimum versions we depend on will mean we'll need to address incompatible diamond dependencies before we might otherwise need to. Not really a big deal, given we can just close the PR if there's no obvious benefit to bumping and we don't feel like dealing with it. One consideration though is that for crossplane-runtime (i.e. a library) we'll effectively be signing up anyone using it to resolve any conflicts that result from a bunch of potentially unnecessary dependency updates any time they update to a new version of crossplane-runtime.

**Review by negz — APPROVED** [review]

**Comment by phisco** [comment]
@negz I'm not sure I've got your point, but in you example, if `yaml` is an indirect dependency, you won't get a PR to bump it per se, it will only be bumped to the minimum version selected by the MVS algorithm.

The only difference with the current behaviour would be that now we bump only direct dependencies and only in case of vulnerabilities, and there is no way to get those updates on indirect ones and no way to transitively trigger bumps on direct dependencies importing affected indirect ones, while enabling all updates to direct dependencies we'll hopefully get less vulnerabilities on indirect dependencies. This should happen because the indirect dependency will be bumped as soon as the minimum version required by all our up-to-date direct dependencies will be higher than the affected one, which if our direct dependencies are healthy and maintained, should happen pretty fast.

**Comment by HonkingGoose** [comment]
The Renovate docs mention indirect updates in the `gomod` manager section: [^1]

> Indirect updates are disabled by default. To enable them, add a package rule such as:
>
> 
> ```json
> {
>   "packageRules": [
>     {
>       "matchManagers": ["gomod"],
>       "matchDepTypes": ["indirect"],
>       "enabled": true
>     }
>   ]
> }
> ```

I also found a related issue:

- https://github.com/renovatebot/renovate/issues/12999

[^1]: [Renovate docs, `gomod` manager, Post-update options](https://docs.renovatebot.com/modules/manager/gomod/#post-update-options)

**Comment by phisco** [comment]
@HonkingGoose yes, but as per the discussion [here](https://github.com/renovatebot/renovate/discussions/22505#discussioncomment-6040777), it's not possible to enable indirect update only on security vulnerabilities

**Comment by negz** [comment]
> I'm not sure I've got your point

@phisco My understanding is that with this change in place, Renovate will bump all Go dependencies whenever a new version of a dependency is available. So assuming we merge all the Renovate PRs, the versions listed in `go.mod` will always be the latest possible versions. Is that right?

The big comment I wrote earlier was really just me thinking out loud to try understand how that relates to "idiomatic" Go version selection. My conclusion was they seem pretty compatible. Go obviously still uses MVS at build-time, it's just that we'll always be telling it the minimum allowed version is the latest available version, even if there were no _known_ security issues with an older version, and even if there's no features we or any other dependency needs in the latest version.

The only downside I could see is that proactively updating all our dependencies means we also have to proactively resolve new dependency conflicts if and when they appear. The good thing is we'd be doing this in order to get a Renovate PR to merge, not in order to unbreak our build.

So TL;DR assuming all this works like I think it works I feel good about merging. Should I go ahead and do that?

**Comment by phisco** [comment]
> So assuming we merge all the Renovate PRs, the versions listed in go.mod will always be the latest possible versions. Is that right?

@negz this will be true only for direct dependencies, indirect ones will only be updated if the MVS algorithm decides it's time to do so because all direct dependencies require an higher minimum version.

And yes, please, you can merge this 🙏

**Pull request merged**